### PR TITLE
Add function to explicitly mark forward methods in Fabric

### LINKED ISSUE #19495: _FabricModule breaks the Huggingface `model.generate()` method
### Bug description

Due to the strict prevention of bypassing model `forward()` call in PR #18819, it breaks the Huggingface generation models, which inherit [GenerationMixin](https://huggingface.co/docs/transformers/en/main_classes/text_generation#transformers.GenerationMixin) class having `generate()` method.

Since `generate()` method is not to call `backward()`, the methd should not be prevented.

### What version are you seeing the problem on?

v2.1, v2.2, master

### How to reproduce the bug

```python
model = AutoModelForCausalLM.from_pretrained("gpt2")
model = Fabric(accelerator="cuda", strategy="ddp").setup_module(model)
model.generate(**some_inputs)
```

### Error messages and logs

```
RuntimeError: You are calling the method `GPT2LMHeadModel.generate()` from outside the model. This will bypass the wrapper from the strategy and result in incorrect behavior in `.backward()`. You should pass your inputs through `forward()`.
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

### More info

_No response_

cc @carmocca @justusschock @awaelchli

### LINKED ISSUE #19608: Allow custom module inference entry methods with Fabric
### Description & Motivation

Let's say you want to create custom base module class `MyModelBase` to use with custom fabric training loop and some particular implementation `MyGPT`.
```python
import torch
from typing import Any

class MyModelBase(torch.nn.Module):
    cfg: dict

    def __init__(self, cfg: dict):
        super().__init__()
        self.cfg = cfg

    def forward(self, *args, **kwargs):
        raise NotImplementedError()

    def training_step(self, batch) -> torch.Tensor:
        raise NotImplementedError()

    def evaluation_step(self, batch) -> dict[str, Any]:
        loss = self.training_step(batch)
        return {'loss': loss}


class MyGPT(MyModelBase):

    def __init__(self, cfg: dict):
        super().__init__(cfg)
        ...

    def forward(self, *args, **kwargs):
        ...

    def training_step(self, batch) -> torch.Tensor:
        ...

    def evaluation_step(self, batch) -> dict[str, Any]:
        ...
```

Attempting to train `MyGPT` will work fine, until training loop hits `evaluation_step` and fabric raises `RuntimeError`.
```
RuntimeError: You are calling the method `MyGPT.evaluation_step()` from outside the model.
This will bypass the wrapper from the strategy and result in incorrect behavior in `.backward()`.
You should pass your inputs through `forward()`.
```

The error will originate from [here](https://github.com/Lightning-AI/pytorch-lightning/blob/096b063d6eeb41567409f4a6b9bac6f5af28ed93/src/lightning/fabric/wrappers.py#L212).

However something doesn't seem quite right.
Only `evaluation_step` triggers this error while `training_step` works properly even though it's definitely not `forward`.
This happens because fabric internally already has [logic](https://github.com/Lightning-AI/pytorch-lightning/blob/096b063d6eeb41567409f4a6b9bac6f5af28ed93/src/lightning/fabric/wrappers.py#L172) for handling non-forward methods but only uses it for undocumented feature, handling `LightningModule`'s methods.

### This logic should be exposed in some form to allow arbitrary inference-entry methods.
For ex. as decorator:
```python
import torch
from typing import Any
from lightning.fabric import inference_entry

class MyModelBase(torch.nn.Module):
    cfg: dict

    def __init__(self, cfg: dict):
        super().__init__()
        self.cfg = cfg

    def forward(self, *args, **kwargs):
        raise NotImplementedError()

    @inference_entry
    def training_step(self, batch) -> torch.Tensor:
        raise NotImplementedError()

    @inference_entry
    def evaluation_step(self, batch) -> dict[str, Any]:
        loss = self.training_step(batch)
        return {'loss': loss}


class MyGPT(MyModelBase):

    def __init__(self, cfg: dict):
        super().__init__(cfg)
        ...

    def forward(self, *args, **kwargs):
        ...

    @inference_entry
    def training_step(self, batch) -> torch.Tensor:
        ...

    @inference_entry
    def evaluation_step(self, batch) -> dict[str, Any]:
        ...
```

### Pitch

_No response_

### Alternatives

_No response_

### Additional context

_No response_

cc @borda @carmocca @justusschock @awaelchli

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #19608
Fixes #19495

Adds a new method `mark_forward_method` to the wrapper class:

```py
model = fabric.setup(model)
model.mark_forward_method(model.generate)

# Now can use `model.generate()` in DDP/FSDP without issues:
prediction = model.generate(input)
```

See the added docs page and linked issues for why this is important.

----
📚 Documentation preview 📚: https://pytorch-lightning--19690.org.readthedocs.build/en/19690/

cc @borda @carmocca @justusschock @awaelchli

### CODE REVIEW

File: src/lightning/fabric/wrappers.py [hunk 1]
```diff
@@ -137,6 +139,7 @@ def module(self) -> nn.Module:
     def forward(self, *args: Any, **kwargs: Any) -> Any:
         """Casts all inputs to the right precision and handles autocast for operations in the module forward method."""
         precision = self._strategy.precision
+
```
Comment (awaelchli) [inline src/lightning/fabric/wrappers.py:142]:
```suggestion
```

File: src/lightning/fabric/wrappers.py [hunk 2]
```diff
@@ -169,6 +171,18 @@ def load_state_dict(  # type: ignore[override]
     ) -> _IncompatibleKeys:
         return self._original_module.load_state_dict(state_dict=state_dict, strict=strict, **kwargs)
 
+    def mark_forward_method(self, method: Union[MethodType, str]) -> None:
+        """Mark a method as a 'forward' method to prevent it bypassing the strategy wrapper (e.g., DDP)."""
+        name = method if isinstance(method, str) else method.__name__
```
Comment (carmocca) [inline src/lightning/fabric/wrappers.py:176]: edge case: Can you add a test for passing a function with the same name as an existing model method?

**Review by carmocca — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: docs/source-fabric/api/wrappers.rst [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+----
+
+
+************************************************
+Using methods other than forward for computation
+************************************************
+
+PyTorch's ``nn.Modules`` have a special contract you need to follow when using them for training: Your forward computation has to be defined in the **forward** method and you should call this forward method directly.
+But sometimes your model may need to define different flavors of forward, like in this example below where the regular forward is used for training, but the `generate` method does something slightly different for inference:
```
Comment (rasbt) [inline docs/source-fabric/api/wrappers.rst:94]:
```suggestion
PyTorch's ``nn.Modules`` have a special contract you need to follow when using them for training: Your forward computation has to be defined in the **`forward`** method and you should call this forward method directly.
But sometimes your model may need to define different flavors of `forward`, like in this example below where the regular `forward` is used for training, but the `generate` method does something slightly different for inference:
```

**Review by lantiga — APPROVED** [review]
LGTM
